File: rate-limit-callback-fires-once/python/rate_limit_guard.py
"""Read BigCommerce's live rate limit headers on every request instead of a one-shot callback.

BigCommerce enforces a sliding request quota per store (default 150 requests per
30,000 ms window for OAuth apps) and reports the live state on every response
through X-Rate-Limit-Requests-Left, X-Rate-Limit-Requests-Quota,
X-Rate-Limit-Time-Window-Ms, and X-Rate-Limit-Time-Reset-Ms. There is no
server-side webhook or push callback for rate limiting, it is purely response
header driven. Client libraries such as bigcommerce-api-python wire a
callback_function into the client once, at construction time, and their
internal "requests remaining" counter is only updated inside their own request
loop rather than re-read from the live headers on every call, so the callback
fires a single time instead of on every request that crosses the threshold.
The script then free runs on stale internal state and keeps colliding with the
real quota, hitting repeated 429 Too Many Requests responses.

This helper reads the four headers off every response and decides, fresh each
time, whether to sleep before the next call. If DRY_RUN=true it only replays a
log of historical responses and prints the computed sleep durations without
making live calls. If DRY_RUN=false it applies the throttling in the live
request loop.

Guide: https://www.allanninal.dev/bigcommerce/rate-limit-callback-fires-once/
"""
import os
import time
import logging
from typing import Optional, Tuple

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("rate_limit_guard")

STORE_HASH = os.environ.get("BIGCOMMERCE_STORE_HASH", "example_hash")
ACCESS_TOKEN = os.environ.get("BIGCOMMERCE_ACCESS_TOKEN", "bc_dummy")
API_BASE = f"https://api.bigcommerce.com/stores/{STORE_HASH}/v3"
MIN_REQUESTS_REMAINING = int(os.environ.get("MIN_REQUESTS_REMAINING", "10"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

HEADERS = {
    "X-Auth-Token": ACCESS_TOKEN,
    "Content-Type": "application/json",
    "Accept": "application/json",
}


def should_throttle(
    requests_left: Optional[int],
    time_reset_ms: Optional[int],
    min_requests_remaining: int = MIN_REQUESTS_REMAINING,
    status_code: int = 200,
) -> Tuple[bool, int]:
    """Pure decision. No network, no side effects.

    Returns (True, time_reset_ms) if status_code == 429 or requests_left is
    missing/negative/<= min_requests_remaining, meaning the caller must sleep
    time_reset_ms before its next request. Otherwise returns (False, 0).
    Missing or invalid header values fail safe toward throttling.
    """
    safe_reset_ms = time_reset_ms if isinstance(time_reset_ms, int) and time_reset_ms > 0 else 0

    if status_code == 429:
        return True, safe_reset_ms

    if requests_left is None or requests_left <= min_requests_remaining:
        return True, safe_reset_ms

    return False, 0


def _parse_rate_headers(response) -> dict:
    def _int_or_none(value):
        try:
            return int(value)
        except (TypeError, ValueError):
            return None

    return {
        "requests_left": _int_or_none(response.headers.get("X-Rate-Limit-Requests-Left")),
        "requests_quota": _int_or_none(response.headers.get("X-Rate-Limit-Requests-Quota")),
        "window_ms": _int_or_none(response.headers.get("X-Rate-Limit-Time-Window-Ms")) or 0,
        "reset_ms": _int_or_none(response.headers.get("X-Rate-Limit-Time-Reset-Ms")) or 0,
        "status_code": response.status_code,
    }


def bc_get(path, params=None):
    response = requests.get(f"{API_BASE}{path}", headers=HEADERS, params=params or {}, timeout=30)
    rate = _parse_rate_headers(response)
    return response, rate


def replay_dry_run(historical_responses):
    """Simulate throttle decisions against a replayed log, no live calls."""
    for entry in historical_responses:
        throttle, wait_ms = should_throttle(
            entry.get("requests_left"),
            entry.get("reset_ms", 0),
            MIN_REQUESTS_REMAINING,
            entry.get("status_code", 200),
        )
        log.info(
            "timestamp=%s requests_left=%s reset_ms=%s status_code=%s throttle=%s wait_ms=%s",
            entry.get("timestamp"), entry.get("requests_left"), entry.get("reset_ms"),
            entry.get("status_code", 200), throttle, wait_ms if throttle else 0,
        )


def run(paths=None):
    paths = paths or ["/catalog/products"]

    if DRY_RUN:
        log.info("DRY_RUN=true, replaying without live calls is expected; pass a historical log to replay_dry_run().")
        return

    for path in paths:
        response, rate = bc_get(path)
        throttle, wait_ms = should_throttle(
            rate["requests_left"], rate["reset_ms"], MIN_REQUESTS_REMAINING, rate["status_code"]
        )
        log.info(
            "path=%s status_code=%s requests_left=%s reset_ms=%s throttle=%s",
            path, rate["status_code"], rate["requests_left"], rate["reset_ms"], throttle,
        )
        if throttle and wait_ms > 0:
            time.sleep(wait_ms / 1000)


if __name__ == "__main__":
    run()

File: rate-limit-callback-fires-once/node/rate-limit-guard.js
/**
 * Read BigCommerce's live rate limit headers on every request instead of a one-shot callback.
 *
 * BigCommerce enforces a sliding request quota per store (default 150 requests per
 * 30,000 ms window for OAuth apps) and reports the live state on every response
 * through X-Rate-Limit-Requests-Left, X-Rate-Limit-Requests-Quota,
 * X-Rate-Limit-Time-Window-Ms, and X-Rate-Limit-Time-Reset-Ms. There is no
 * server-side webhook or push callback for rate limiting, it is purely response
 * header driven. Client libraries wire a callback into the client once, at
 * construction time, and their internal "requests remaining" counter is only
 * updated inside their own request loop rather than re-read from the live
 * headers on every call, so the callback fires a single time instead of on
 * every request that crosses the threshold. The script then free runs on
 * stale internal state and keeps colliding with the real quota, hitting
 * repeated 429 Too Many Requests responses.
 *
 * This helper reads the four headers off every response and decides, fresh
 * each time, whether to sleep before the next call. If DRY_RUN=true it only
 * replays a log of historical responses and prints the computed sleep
 * durations without making live calls. If DRY_RUN=false it applies the
 * throttling in the live request loop.
 *
 * Guide: https://www.allanninal.dev/bigcommerce/rate-limit-callback-fires-once/
 */
import { pathToFileURL } from "node:url";

const STORE_HASH = process.env.BIGCOMMERCE_STORE_HASH || "example_hash";
const ACCESS_TOKEN = process.env.BIGCOMMERCE_ACCESS_TOKEN || "bc_dummy";
const API_BASE = `https://api.bigcommerce.com/stores/${STORE_HASH}/v3`;
const MIN_REQUESTS_REMAINING = Number(process.env.MIN_REQUESTS_REMAINING || 10);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const HEADERS = {
  "X-Auth-Token": ACCESS_TOKEN,
  "Content-Type": "application/json",
  Accept: "application/json",
};

/**
 * Pure decision. No network, no side effects.
 *
 * Returns [true, timeResetMs] if statusCode === 429 or requestsLeft is
 * missing/negative/<= minRequestsRemaining, meaning the caller must sleep
 * timeResetMs before its next request. Otherwise returns [false, 0].
 * Missing or invalid header values fail safe toward throttling.
 */
export function shouldThrottle(requestsLeft, timeResetMs, minRequestsRemaining = MIN_REQUESTS_REMAINING, statusCode = 200) {
  const safeResetMs = Number.isFinite(timeResetMs) && timeResetMs > 0 ? timeResetMs : 0;

  if (statusCode === 429) return [true, safeResetMs];

  if (requestsLeft == null || requestsLeft <= minRequestsRemaining) return [true, safeResetMs];

  return [false, 0];
}

function parseRateHeaders(res) {
  const toIntOrNull = (value) => {
    const parsed = Number.parseInt(value, 10);
    return Number.isFinite(parsed) ? parsed : null;
  };

  return {
    requestsLeft: toIntOrNull(res.headers.get("X-Rate-Limit-Requests-Left")),
    requestsQuota: toIntOrNull(res.headers.get("X-Rate-Limit-Requests-Quota")),
    windowMs: toIntOrNull(res.headers.get("X-Rate-Limit-Time-Window-Ms")) || 0,
    resetMs: toIntOrNull(res.headers.get("X-Rate-Limit-Time-Reset-Ms")) || 0,
    statusCode: res.status,
  };
}

async function bcGet(path, params = {}) {
  const url = new URL(`${API_BASE}${path}`);
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null) url.searchParams.set(key, value);
  }
  const res = await fetch(url, { headers: HEADERS });
  const rate = parseRateHeaders(res);
  return { res, rate };
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Simulate throttle decisions against a replayed log, no live calls. */
export function replayDryRun(historicalResponses) {
  for (const entry of historicalResponses) {
    const [throttle, waitMs] = shouldThrottle(
      entry.requestsLeft, entry.resetMs || 0, MIN_REQUESTS_REMAINING, entry.statusCode ?? 200
    );
    console.log(
      `timestamp=${entry.timestamp} requests_left=${entry.requestsLeft} reset_ms=${entry.resetMs} ` +
      `status_code=${entry.statusCode ?? 200} throttle=${throttle} wait_ms=${throttle ? waitMs : 0}`
    );
  }
}

export async function run(paths = ["/catalog/products"]) {
  if (DRY_RUN) {
    console.log("DRY_RUN=true, replaying without live calls is expected; pass a historical log to replayDryRun().");
    return;
  }

  for (const path of paths) {
    const { rate } = await bcGet(path);
    const [throttle, waitMs] = shouldThrottle(rate.requestsLeft, rate.resetMs, MIN_REQUESTS_REMAINING, rate.statusCode);
    console.log(
      `path=${path} status_code=${rate.statusCode} requests_left=${rate.requestsLeft} reset_ms=${rate.resetMs} throttle=${throttle}`
    );
    if (throttle && waitMs > 0) await sleep(waitMs);
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
